### PR TITLE
WT-11669 Create new log record for backup ids.

### PR DESCRIPTION
**Pull request by sueloverso**
Restore backup ids on recovery.

### CODE REVIEW

**Comment by sueloverso** [comment]
@pmacko86 this working so far but isn't ready for merge until the reproducer is merged and heavily tested with this.

File: src/log/log_sys.c
Comment (pmacko86) [inline src/log/log_sys.c:45]: As this code is logging only existing backup IDs, does it also handle the case when a valid backup ID is removed (and is not replaced by a new backup ID)? For example, this would be the case of using `force_stop`.

**Review by pmacko86 — COMMENTED** [review]
This is a nice start to the solution, but I am not sure if this would actually fix the problem in its entirety.

As I understand this PR, it logs the valid backup IDs at the beginning of the checkpoint, but it does not actually log removal of any backup IDs. This could create two problems:
* In the case of `force_stop`, the removal of a backup ID is not logged, so if WiredTiger crashes just before the turtle file is updated, the force-stopped backup IDs may reappear. So this might not solve the second problem (the one with force-stop) described in WT-11669.
* Because we are now logging new backup IDs before the corresponding changes to the metadata table, if we crash after this log record is written but before the metadata table is updated, we will have a new backup ID after recovery that the tables do not know about. I'm afraid that might introduce another problem backup inconsistency.

The way I understand the issue, here are two ways that I believe would solve the problem in its entirety:
* If we would like to log backup IDs before the checkpoint starts, log only the backup IDs that are being removed—whether it is by being replaced by new backup IDs (as in the case of regular incremental backups) or removed entirely (as with force-stop). In that way, old backup IDs will be removed just before they disappear from the metadata table, so they will not reappear after a crash. New backup IDs will then appear only after the turtle file is updated at the end of the checkpoint, i.e., only after they are recorded in the metadata table, giving us the right semantics. (This is probably the easier way to do this than the next option.)
* If we would like to log all backup IDs, we need to log both the removal and the addition of backup IDs at the same time, and we would have to do this atomically with updates to the metadata table. That is, this would have to be included in the `commit` record that corresponds to the metadata table. The reason is that old backup IDs must removed before or at the same time as the metadata table is updated (as described in the ticket), and the new backup IDs can appear only after or at the same time as the metadata table is updated (as explained above)—so if we are doing removal of old IDs and the addition of new IDs in the same log records, it would have to be done atomically with the metadata table update.

**Comment by sueloverso** [comment]
@pmacko86 I'd like you to take another look at this as I believe the changes related to backup are complete. I've been running `test_wt12015_backup_corruption` and `timestamp_abort -B -I 3` and `test_incr_backup` in a loop for a while and all are passing.

I'm leaving it as a Draft at the moment while I think more about the log version change and the implications of that. But the backup piece can be reviewed.

**Comment by sueloverso** [comment]
@pmacko86 and @keitharnoldsmith although this is ready for review and passing tests, there is a bit of refactoring I want to do. Hold off for a day.

**Comment by sueloverso** [comment]
@pmacko86 and @keitharnoldsmith I've completed the changes I had in mind so this is ready for review.

File: src/txn/txn_recover.c
Comment (pmacko86) [inline src/txn/txn_recover.c:36]: This does not seem to be used once set.

File: src/txn/txn_recover.c
Comment (pmacko86) [inline src/txn/txn_recover.c:183]: This does not seem to be used once set, but if we are keeping it, this assignment does not seem right.

**Review by pmacko86 — APPROVED** [review]
LGTM with minor comments.

File: src/txn/txn_recover.c
Comment (sueloverso) [reply src/txn/txn_recover.c:183]: You're right on both counts. This must have been left over from something else I tried. I will remove it.

**Comment by sueloverso** [comment]
I also just added a test of the new debug_info setting to a python test. I rerequested a review from @keitharnoldsmith so someone eyeballs it (it is just a few new lines in an existing test). @pmacko86 you only need to look if you're interested.

**Review by keitharnoldsmith — APPROVED** [review]
Thanks Sue. LGTM! 

There is one minor suggestion, below. But nothing that I need to re-review.

I do have one general question. If I understand the changes in `__wt_txn_recover()`, this adds a pass through the entire log during backup restore in order to find any backup ID log records. Could this have enough effect on restore time that we should worry about it? I'm not sure what we would do about it, but it might at least merit a ticket to point out and track it?

**Comment by pmacko86** [comment]
@sueloverso, one more question: As this adds a new log entry type, do we need to worry about upgrade/downgrade compatibility? Do we need any testing around that?

File: src/txn/txn_recover.c
Comment (sueloverso) [reply src/txn/txn_recover.c:191]: I like that better. Thanks.

**Comment by sueloverso** [comment]
I looked at the history of other operation changes and found consistently that they did not change the log version. In all cases (including this one) the new operation did not affect running and recovery if the same log was used in a downgraded system. The only change/issue in a downgraded system is running `printlog`.

Earlier changes to add new or modify existing operations are in WT-3440, WT-4712, WT-5766 and WT-6252.

**Comment by sueloverso** [comment]
evergreen merge

**Pull request merged**
